src/coreclr/jit/emitriscv64.cpp

```diff
@@ -2118,8 +2118,8 @@ void emitter::emitJumpDistBind()
 unsigned emitter::emitOutput_Instr(BYTE* dst, code_t code) const
 {
     assert(dst != nullptr);
-    assert(sizeof(code_t) == 4);
-    memcpy(dst + writeableOffset, &code, sizeof(code_t));
+    static_assert(sizeof(code_t) == 4, "code_t must be 4 bytes");
+    *(code_t*)(dst + writeableOffset) = code;
```

Comment:
The buffer is not guaranteed to be aligned (for example in prejit it is just the result of a normal byte array allocation). Are unaligned writes on RV64 allowed?

Reply:
VisionFive 2 allows it and I think most 64-bit implementations do. The [RVA20U64 profile mandates it](https://github.com/riscv/riscv-profiles/blob/54717ad320c37f27d235483f23e3f0a805235c37/profiles.adoc?plain=1#L529-L538) but currently we compile for RV64GC ISA which does allow trapping and I don't see `-mno-strict-align` anywhere so let's play it safe. Thank you for noticing.

This could be \#ifdef'ed in the future when we get the [recommended macros in clang](https://github.com/llvm/llvm-project/pull/65756), I don't see them in 17.0.6 yet.

Reply:
> RV64GC ISA which does allow trapping

RV64GC and RVA20U64 are the same thing. RVA20 is just a new, extensible, name for RV64GC. RISC-V has always required User-mode misaligned access to work. It *may* trap, but it must be trap-and-emulate and work in the end (but slowly).  When RVA20U64 says "Even though mandated, misaligned loads and stores might execute extremely slowly." what they mean is it may be trap-and-emulate. EXTREMELY slowly. There is no change in this from RV64GC.

> The buffer is not guaranteed to be aligned (for example in prejit it is just the result of a normal byte array allocation).

It would be very unusual in any memory allocator for any object to not be aligned to at least 4 bytes, if not 16. Is this really likely?

Reply:
During prejit (where code is being emitted into an object file in the end) this is just the result of a normal byte array allocation coming from the EE side. The buffer is likely to be aligned, and in practice it will be, but it is not a guarantee provided by the JIT-EE interface. If you want to allocate a temporary buffer yourself and guarantee its alignment, then you can do that.

> It seems to me this is pretty performance-critical code, and really doesn't want to suffer calling memcpy() -- which the compiler can't optimise out as it knows casting a BYTE* to a code_t* does not make an aligned pointer.

Writing the code to the buffer is a drop in the bucket compared to everything else that goes on during codegen. Each IR node that resulted in the instructions written are orders of magnitudes larger than the corresponding final bytes written to the buffer.

The compiler is perfectly capable of optimizing out memcpy calls with constant size. That is optimized by all compilers I know of. @tomeksowi pointed out above that it is turned into some sequence of sb + srl on RISCV. It does not turn into a call.

Reply:
> @tomeksowi pointed out above that it is turned into some sequence of sb + srli on RISCV

[Yes](https://godbolt.org/z/P55aeveMv), and given we will support compressed instructions eventually, in the long run the improvement will be shaving off 4 instructions (2x sh + 1x srli vs. 4x sb + 3x srli now), that's if we guarantee 2-byte alignment ourselves. IMO juice not worth the squeeze.